### PR TITLE
[SPARK-48403][SQL] Fix Lower & Upper expressions for UTF8_BINARY_LCASE & ICU collations

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
String lowercase/uppercase conversion in UTF8_BINARY_LCASE now works using ICU default locale, similar to how other ICU collations currently work in Spark.

### Why are the changes needed?
All collations apart from UTF8_BINARY should use the same interface (UCharacter) that utilizes ICU toLowerCase/toUpperCase implementation, rather than mixing JVM & ICU implementations.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing unit tests and e2e sql tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 1]
```diff
@@ -135,22 +135,90 @@ public static UTF8String lowercaseReplace(final UTF8String src, final UTF8String
...
+   */
+  public static UTF8String toTitleCase(final UTF8String target, final int collationId) {
+    return UTF8String.fromString(toTitleCase(target.toString(), collationId));
+  }
   public static String toTitleCase(final String target, final int collationId) {
     ULocale locale = CollationFactory.fetchCollation(collationId)
       .collator.getLocale(ULocale.ACTUAL_LOCALE);
-    return UCharacter.toTitleCase(locale, target, BreakIterator.getWordInstance(locale));
+    BreakIterator wordIterator = BreakIterator.getWordInstance(locale);
+    return UCharacter.toTitleCase(locale, target, wordIterator);
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:221]: Why are we making this change?

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 2]
```diff
@@ -135,22 +135,90 @@ public static UTF8String lowercaseReplace(final UTF8String src, final UTF8String
...
+    return UTF8String.fromString(toLowerCase(target.toString(), collationId));
+  }
   public static String toLowerCase(final String target, final int collationId) {
     ULocale locale = CollationFactory.fetchCollation(collationId)
       .collator.getLocale(ULocale.ACTUAL_LOCALE);
     return UCharacter.toLowerCase(locale, target);
   }
 
+  /**
+   * Convert the input string to lowercase using the ICU root locale rules.
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:195]:
```suggestion
   * Convert the input string to titlecase using the ICU root locale rules.
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 3]
```diff
@@ -135,22 +135,90 @@ public static UTF8String lowercaseReplace(final UTF8String src, final UTF8String
...
+   *
+   * @param target the input string
+   * @return the lowercase string
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:198]:
```suggestion
   * @return the titlecase string
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 4]
```diff
@@ -135,22 +135,90 @@ public static UTF8String lowercaseReplace(final UTF8String src, final UTF8String
...
+   */
+  public static UTF8String toTitleCase(final UTF8String target) {
+    return UTF8String.fromString(toTitleCase(target.toString()));
+  }
+  public static String toTitleCase(final String target) {
+    BreakIterator wordIterator = BreakIterator.getWordInstance();
+    return UCharacter.toTitleCase(target, wordIterator);
+  }
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:206]: I suggest we do this in a separate PR and probably think a bit more about the intended behavior.
If this is meant to be used for implementing initcap, notice that it will break the string not all at ASCII space, but also a lot of other characters (and it is context sensitive).

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 5]
```diff
@@ -208,87 +208,99 @@ public static boolean execICU(final UTF8String l, final UTF8String r,
...
+    }
     public static UTF8String execICU(final UTF8String v, final int collationId) {
-      return UTF8String.fromString(
-              CollationAwareUTF8String.toTitleCase(
-                      CollationAwareUTF8String.toLowerCase(
-                              v.toString(),
-                              collationId
-                      ),
-                      collationId));
+      return CollationAwareUTF8String.toTitleCase(v, collationId);
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:303]: As I mention in another comment: this is going to title case letters not just after the ASCII space, but a whole lot more. This is something to be discussed.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 6]
```diff
@@ -208,87 +208,99 @@ public static boolean execICU(final UTF8String l, final UTF8String r,
...
+        return String.format(expr + "Lowercase(%s)", v);
       } else {
         return String.format(expr + "ICU(%s, %d)", v, collationId);
       }
     }
-
     public static UTF8String execUTF8(final UTF8String v) {
       return v.toLowerCase().toTitleCase();
     }
-
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:283]: Are we using this to implement initcap? If so, this is a breaking change (in many ways).

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java (hunk 4 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:206]: agreed, removing

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 5 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:303]: separating this out

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java (hunk 2 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:195]: (removing this anyway)

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java (hunk 1 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:221]: (removing this anyway)

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 6 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java]: so this code is now removed, but you're referring to the UTF8_BINARY_LCASE (default UTF8String / pre-collation Spark) implementation:

```
public static UTF8String execUTF8(final UTF8String v) {
  return v.toLowerCase().toTitleCase();
}
```

so essentially this is how it was done in Spark before introducing collation awareness for InitCap (i.e. `public static class InitCap` in `CollationSupport`), so no breaking changes

here's the old code for reference (from `UTF8String`):

```
  override def nullSafeEval(string: Any): Any = {
    // scalastyle:off caselocale
    string.asInstanceOf[UTF8String].toLowerCase.toTitleCase
    // scalastyle:on caselocale
```

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java]: (btw this is no longer relevant for the current PR)

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 7]
```diff
@@ -284,12 +284,56 @@ public static UTF8String lowercaseReplace(final UTF8String src, final UTF8String
...
 
+  /**
+   * Convert the input string to uppercase using the ICU root locale rules.
+   *
+   * @param target the input string
+   * @return the uppercase string
+   */
+  public static UTF8String toUpperCase(final UTF8String target) {
+    return UTF8String.fromString(toUpperCase(target.toString()));
+  }
```
Comment (dbatomic) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:295]: empty line here

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 8]
```diff
@@ -284,12 +284,56 @@ public static UTF8String lowercaseReplace(final UTF8String src, final UTF8String
...
+
+  /**
+   * Convert the input string to uppercase using the specified ICU collation rules.
+   *
+   * @param target the input string
+   * @return the uppercase string
+   */
+  public static UTF8String toUpperCase(final UTF8String target, final int collationId) {
+    return UTF8String.fromString(toUpperCase(target.toString(), collationId));
+  }
```
Comment (dbatomic) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:308]: empty line here

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java [hunk 9]
```diff
@@ -494,10 +494,10 @@ public void testUpper() throws SparkException {
...
     assertUpper("äbćδe", "UNICODE_CI", "ÄBĆΔE");
     // Case-variable character length
-    assertUpper("i̇o", "UTF8_BINARY","İO");
-    assertUpper("i̇o", "UTF8_BINARY_LCASE","İO");
-    assertUpper("i̇o", "UNICODE","İO");
-    assertUpper("i̇o", "UNICODE_CI","İO");
+    assertUpper("i\u0307o", "UTF8_BINARY","I\u0307O");
+    assertUpper("i\u0307o", "UTF8_BINARY_LCASE","I\u0307O");
+    assertUpper("i\u0307o", "UNICODE","I\u0307O");
+    assertUpper("i\u0307o", "UNICODE_CI","I\u0307O");
```
Comment (mkaravel) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:500]: Please add some _interesting_ test cases for uppercasing, notably strings that involve ligatures or characters that map to multiple characters when uppercased:
```scala
assertUpper("ß ﬁ ﬃ ﬀ ﬆ ῗ", "UTF8_BINARY","SS FI FFI FF ST \u0399\u0308\u0342");
assertUpper("ß ﬁ ﬃ ﬀ ﬆ ῗ", "UTF8_BINARY_LCASE","SS FI FFI FF ST \u0399\u0308\u0342");
assertUpper("ß ﬁ ﬃ ﬀ ﬆ ῗ", "UNICODE","SS FI FFI FF ST \u0399\u0308\u0342");
assertUpper("ß ﬁ ﬃ ﬀ ﬆ ῗ", "UNICODE","SS FI FFI FF ST \u0399\u0308\u0342");
```

**Review by mkaravel — APPROVED** [review]
LGTM. Please add a few more interesting test cases for uppercasing in this PR or a follow up one.

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java (hunk 9 above)
Comment (uros-db) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:500]: added

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
